### PR TITLE
[11.x] Improved Performance for Testing with In-Memory Databases

### PR DESCRIPTION
**Pull request by AJenbo**
This pull request aims to enhance the performance and developer experience when running tests with an in-memory database in Laravel. By keeping a reference to the PDO object and restoring it for connections, we enable transaction-based database refresh, resulting in a significant performance boost. Test suites that used to take minutes to run are now completed in just seconds, providing a much smoother testing process.

image

I believe this change is non-breaking as it replicates the same database operations that would occur with a persisted database configuration. However, if there are any concerns, please let me know, and I can rebase and retarget the pull request on the master branch.

Please feel free to provide any feedback or preferences regarding the implementation or coding style. Your input is greatly appreciated.

Thank you.

### CODE REVIEW

**Comment by JurianArie** [comment]
I like the idea, and I think it could work, but there is a bug in your implementation. Copy pasting the trait breaks my entire test suite.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]
Probably 11.x is the best place for this, as this could break things for people doing funky things such as DDL in their tests.

**Comment by AJenbo** [comment]
@GrahamCampbell 

Thank you for the feedback. I have rebased on `master` and set it as the target branch as well as updated the title to reflect that this is targeting Larvel 11.x.

@JurianArie Can you please be more specific here is what I just did:

- Crate a Laravel 11 skeleton app
- Copy .env
- Generate keys
- Set DB to `sqlite` `:memory:` in .env
- Delete the unit test
- Duplicate the feature test
- Add `sleep(5)` to the create_users migration
- Run test
- Swap `RefreshDatabase` to the one from this PR
- Run test

Result:
image

**Comment by AJenbo** [comment]
For a bit of context this is a technique I have used on several project (that where worked on by multiple developers) over the past 4 years so it's been pretty well tested prior to this PR.

**Comment by JurianArie** [comment]
@AJenbo I copy pasted the trait from this PR into a project that uses a in memory DB for testing. It looks like the database never gets migrated.

<details>
<summary>Error output</summary>



```
phpunit --stop-on-error
PHPUnit 10.2.6 by Sebastian Bergmann and contributors.

Runtime:       PHP 8.1.15
Configuration: [SNIP]phpunit.xml

............E

Time: 00:04.762, Memory: 62.50 MB

There was 1 error:

1) Tests\Feature\[TESTCASE]
Illuminate\Database\QueryException: SQLSTATE[HY000]: General error: 1 no such table: files (Connection: sqlite, SQL: insert into "files" [SNIP])

vendor/laravel/framework/src/Illuminate/Database/Connection.php:574
vendor/laravel/framework/src/Illuminate/Database/Connection.php:788
vendor/laravel/framework/src/Illuminate/Database/Connection.php:755
vendor/laravel/framework/src/Illuminate/Database/Connection.php:581
vendor/laravel/framework/src/Illuminate/Database/Connection.php:533
vendor/laravel/framework/src/Illuminate/Database/Query/Processors/Processor.php:32
vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:3383
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:1925
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1333
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1298
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1137
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:330
vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php:236
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:339
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:279
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:475
vendor/laravel/framework/src/Illuminate/Collections/Arr.php:558
vendor/laravel/framework/src/Illuminate/Collections/Collection.php:772
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:472
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:425
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:409
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/GuardsAttributes.php:155
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:414
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:382
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:276
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:301
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php:408
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:302
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:232
[TESTCASE]:36
vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:174

Caused by
PDOException: SQLSTATE[HY000]: General error: 1 no such table: files

vendor/laravel/framework/src/Illuminate/Database/Connection.php:574
vendor/laravel/framework/src/Illuminate/Database/Connection.php:788
vendor/laravel/framework/src/Illuminate/Database/Connection.php:755
vendor/laravel/framework/src/Illuminate/Database/Connection.php:581
vendor/laravel/framework/src/Illuminate/Database/Connection.php:533
vendor/laravel/framework/src/Illuminate/Database/Query/Processors/Processor.php:32
vendor/laravel/framework/src/Illuminate/Database/Query/Builder.php:3383
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Builder.php:1925
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1333
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1298
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Model.php:1137
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:330
vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php:236
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:339
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:279
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:475
vendor/laravel/framework/src/Illuminate/Collections/Arr.php:558
vendor/laravel/framework/src/Illuminate/Collections/Collection.php:772
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:472
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:425
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:409
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/GuardsAttributes.php:155
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:414
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:382
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:276
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:301
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Concerns/HasEvents.php:408
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:302
vendor/laravel/framework/src/Illuminate/Database/Eloquent/Factories/Factory.php:232
[TESTCASE]:36
vendor/laravel/framework/src/Illuminate/Foundation/Testing/TestCase.php:174
```

</details>

**Comment by AJenbo** [comment]
Can you please provide a test case or steps to reproduce it.

**Comment by mbardelmeijer** [comment]
I can confirm a good speedup as well on our test suite with these changes 👍 

Without change: `Time: 01:45.759, Memory: 123.23 MB`
With change: `Time: 00:36.129, Memory: 125.00 MB`

PHPUnit 10.2.6 with PHP 8.2.8

**Comment by nshiro** [comment]
I can also confirm an increase in speed.

182 tests
Without change: Time: 2.65s
With change: Time: 1.68s

PHPUnit 10.2.6 (+ with some Pest)
PHP 8.1

@AJenbo
Since the `migrateUsing` function is no longer called, should we remove it?

**Comment by axlon** [comment]
This breaks my testsuite as well, running a single test or test file works fine, but running the whole suite breaks all tests that rely on the database.
I've tested with and without Paratest, both give the same result, which is an error about a non-existent table (implying the migrations weren't run)

Hope this helps.

**Comment by AJenbo** [comment]
@axlon Can you please provide a test case or steps to reproduce it.

**Comment by taylorotwell** [comment]
Please mark as ready for review when cause of breakages has been found and fixed.

**Comment by schonhoff** [comment]
Hi,

I tried out the trait and got the same error, reported earlier. I switched `LazilyRefreshDatabase` to `RefreshDatabase` (your modified trait). No other changes were made. My configuration:

phpunit.xml:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<phpunit xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="https://schema.phpunit.de/10.1/phpunit.xsd"
         bootstrap="vendor/autoload.php"
         cacheResult="true"
         backupGlobals="false"
         colors="true"
         cacheDirectory="coverage/cache"
         backupStaticProperties="false"
>
    <testsuites>
        <testsuite name="Feature">
            <directory suffix="Test.php">./tests/Feature</directory>
        </testsuite>
        <testsuite name="Unit">
            <directory suffix="Test.php">./tests/Unit</directory>
        </testsuite>
    </testsuites>
    <source>
        <include>
            <directory suffix=".php">./app</directory>
        </include>
    </source>
    <php>
        <env name="APP_ENV" value="testing"/>
        <env name="BCRYPT_ROUNDS" value="4"/>
        <env name="CACHE_DRIVER" value="array"/>
        <env name="MAIL_DRIVER" value="array"/>
        <env name="QUEUE_DRIVER" value="sync"/>
        <env name="SESSION_DRIVER" value="array"/>
        <env name="TELESCOPE_ENABLED" value="false"/>
    </php>
</phpunit>
```

.env.testing
```env
[...]
DB_CONNECTION=sqlite_testing
DB_HOST=127.0.0.1
DB_PORT=3306
[...]
```

database.php (config)
```php
[...]
'connections' => [
    'sqlite_testing' => [
        'driver' => 'sqlite',
        'database' => ':memory:',
        'prefix' => '',
        'foreign_key_constraints' => true,
    ],
],
[...]
```

Error in the console after running `php artisan test --stop-on-failure` (this is just an example)
```batch
SQLSTATE[HY000]: General error: 1 no such table: roles (Connection: sqlite_testing, SQL: insert into "roles" [...]
```

I'm using pest version `v2.12.2` as my testing framework with mostly unit tests (want to switch to pest soon).
My app has close to 9000 tests and 40% is using the database. That is why I'm currently working with `LazilyRefreshDatabase` instead `RefreshDatabase`. I'm implementing the trait in the TestCase abstract class.


Hope that helps to narrow down the error a bit more.

**Comment by nunomaduro** [comment]
Results using Vapor is test suite: Apple M1 Pro 16 GB | 851 tests

```
Before this pull request:
./vendor/bin/pest: 231.74s
./vendor/bin/pest --parallel (10 processes): 23.09s

After this pull request:
./vendor/bin/pest: FAILED - SQLSTATE[HY000]: General error: 1 no such table: users...
./vendor/bin/pest --parallel (10 processes):  FAILED - SQLSTATE[HY000]: General error: 1 no such table: users...
```

File: src/Illuminate/Foundation/Testing/RefreshDatabase.php [hunk 1]
```diff
@@ -4,11 +4,19 @@
 
 use Illuminate\Contracts\Console\Kernel;
 use Illuminate\Foundation\Testing\Traits\CanConfigureMigrationCommands;
+use PDO;
```
Comment (AJenbo) [inline src/Illuminate/Foundation/Testing/RefreshDatabase.php:7]:
```suggestion
```

**Comment by AJenbo** [comment]
@nunomaduro sorry I'm not sure I follow. Are you saying that using Vapor in a project will trigger the issue, or that the test suite for the Vaport project displays the issue?
(full disclaimer I have no knolage about what Vapor is other then hearing the name before)

**Comment by nunomaduro** [comment]
@AJenbo Just saying that the changes on this pull request makes the Vapor's test suite to fail.

**Comment by AJenbo** [comment]
@nunomaduro sorry for my ignorance but I'm not sure how I can run this myself to debug the issue.I tried signing up on vapor.laravel.com but I didn't see anything about tests, I also tried running the tests on github.com/laravel/vapor-core but didn't have any issues after swapping the trait. Is it one of the other vapor projects or am I missing some other steps?

**Comment by AJenbo** [comment]
@schonhoff re applied the configurations you posted to my own projects but was still unable to replicate the issue.

I have two suggestions as to what might solve the issue.
1: change line 95, from:
```php
            if ($this->usingInMemoryDatabase() && ! isset(static::$inMemoryConnections[$name])) {
```
to
```php
            if ($this->usingInMemoryDatabase()) {
```
2: At the start of the `restoreInMemoryDatabase()` function on line 56 add:
```php
        if (! RefreshDatabaseState::$migrated) {
            return;
        }
```

If anyone who is encountering issues could try that out or give a few more hints on how to replicate the affected setup I would very much appreciate it :)

**Comment by donnysim** [comment]
@AJenbo I also get the same exception that a table is not found. It doesn't happen if you have only one feature test with `use RefreshDatabase;`, but it does if you have multiple like `Feature1Test`, `Feature2Test`. And the problem seems more like the connection is always recreated between test classes so you cannot use the `:memory:` driver as each class will have a separate Application instance and separate sqlite memory database, which the second time it does not migrate it and it stays empty.

**Comment by AJenbo** [comment]
@donnysim please see my [comment here](https://github.com/laravel/framework/pull/47912#issuecomment-1661126502) where I give give the step by step of setting up a project with multiple feature tests. I'm aware that new applications instances are created between each test, this is the problem this PR addresses by keeping the PDO instance alive and restoring it in to the new instance between tests. What I'm missing is what you might be doing different from the steps that I give for setting up multiple tests.
If you can provide either the extra steps you do, a test project or try out the suggested modifications from [this comment](https://github.com/laravel/framework/pull/47912#issuecomment-1674944353) that would be appreciated.

I have tested it with 4 different projects with hundreds of test and did not encounter any issues there and as you can see there are also comments here from others who did not have the issue. Question is what we are doing differently from each other.

**Comment by AJenbo** [comment]
I managed to track down [a project](https://github.com/tauseedzaman/contact-management-system) that exhibited the described issue and have applied the needed changes to fix it.

Turns out the difference is that it worked when `RefreshDatabase` was used on the abstract `TestCase` but not the individual test classes. Since this creates individual instances of `$inMemoryConnections`  that then go out of sync with `$migrated`.
The solution was simply to move the property to `RefreshDatabaseState` rather then `RefreshDatabase` so that they share scope.

Thanks to everyone that helped test this.

**Comment by axlon** [comment]
@AJenbo can confirm it is now working as expected! 👍

**Comment by schonhoff** [comment]
Hey,

I tried to use your trait after the fix with no changes on my tests (using `LazilyRefreshDatabase` and pest). It worked:

Before:
image


After:
image

This would be after 10 tests runs on the before configuration and after configuration in decrease of ~35%.

**Comment by driesvints** [comment]
@AJenbo Laravel.io's codebase might be a good candidate as well to try these changes: https://github.com/laravelio/laravel.io

**Review by JurianArie — COMMENTED** [review]
I can confirm that it now works!

**Comment by taylorotwell** [comment]
Sorry - it's been a while since I wrote this code. What happens currently (without this PR) if you try to treat SQLite in-memory databases like MySQL databases and let the `refreshTestDatabase` method run in either case? What error is thrown?

**Comment by AJenbo** [comment]
@taylorotwell Just to clarify, this PR isn't focused on fixing an error, but rather on enabling `RefreshDatabase` to deliver the same performance benefits for an in-memory database as it does for disk-based ones.

Without any changes, each test currently spawns its own instance of an SQLite database in memory and runs migrations on it. This results in a significant amount of time being spent on running the same migrations repeatedly between each test. Essentially, `RefreshDatabase` acts similarly to `DatabaseMigrations` when it detects an in-memory SQLite configuration.

If we attempt to treat the in-memory SQLite the same way as MySQL (or a file-based SQLite), we encounter a `SQLSTATE[HY000]: General error: 1 no such table` error for all but the first test that attempts to use the database. This issue arises because the PDO object (and consequently, the in-memory database) on which the migrations were initially run, gets dropped from memory when the application is torn down between tests. As a result, a new instance is spawned where migrations have not been executed.

**Comment by taylorotwell** [comment]
Thanks!

**Comment by mokhosh** [comment]
@AJenbo is there any documentation on how to migrate a Laravel 10 package to adapt to this change?
We're trying to support Laravel 11 on Filament, but most tests are failing, and I ended up here after creating this issue on testbench https://github.com/orchestral/testbench/issues/393

**Comment by AJenbo** [comment]
@mokhosh does this behave differently for you depending  if you use MySQL or SQLite? Was that also the case prior?

You might want to use DatabaseMigrations or DatabaseTruncation if your tests require that the database is reset between each test. See: https://laravel.com/docs/10.x/dusk#resetting-the-database-after-each-test

**Comment by mokhosh** [comment]
@AJenbo Thanks for your reply.

If I don't touch anything except changing the driver to `mysql` I get `Base table or view already exists`

`DatabaseMigrations` reduced failures from almost 100% to about 50%, and I got `General error: 1 no such table: users`.

`DatabaseTruncation` reduced failures to only one test, and that's the first test, but in that test I'm still getting `General error: 1 no such table: users`, so it still seems related to how these traits work, rather than the tests.

**Comment by AJenbo** [comment]
> If I don't touch anything except changing the driver to mysql I get Base table or view already exists

Ok that's a good indication that your tests are not compatible with `RefreshDatabase` in general and only worked in previous versions because the trait was mostly skipped.

I would suggest you switch to `DatabaseTruncation` and make sure that the user test is run before any of your own migrations.

**Pull request merged**